### PR TITLE
Add 6.3 support

### PR DESCRIPTION
**Pull request by johnpmitsch**
This adds an ansible module to check the metadata file in the backup
for a specific version. Because metadata.yml was added in 6.2.z,
the previous check of satellite-answers is kept for older 6.2 backups.
Any backup with metadata.yml will use the metadata file as a check.

Todo:

- [x] test on 6.3 normal backup
- [x] test on 6.3 online backup
- [x] test on 6.2 backup with metadata.yml

### CODE REVIEW

**Comment by johnpmitsch** [comment]
yee-haw

**Comment by johnpmitsch** [comment]
@evgeni If you have a minute to look at this from an ansible + python perspective, that would be awesome :smile:

**Comment by evgeni** [comment]
Is there a reason you do a "check for version" (twice) instead of "get the version"? plus, you're looking at the katello version, you could also look at the satellite rpm?

**Comment by evgeni** [comment]
I think I'd make the module in a way, that it gets the path to a config tarball, and unarchive, read etc all happens in the module, and the return is `[6.3, 6.2, 'unknown']`, and if it's unknown, you can assume 6.1 later.

**Comment by evgeni** [comment]
oh, the unarchive is not for the metadata, me--

**Comment by johnpmitsch** [comment]
@evgeni Do you think I should check both satellite rpm + katello plugin version? Maybe I should just check the satellite rpm as that is the version I need in the first place. 

My only be concern is grabbing another satellite rpm (like satellite-cli) by mistake, but I should be able to put together some regex to match :)

**Comment by johnpmitsch** [comment]
> return is [6.3, 6.2, 'unknown'], and if it's unknown, you can assume 6.1 later

I like returning a string with the version instead of the t/f check for version. 6.1 backups don't have metadata.yaml files (and older 6.2 z streams) so I would probably error if the module doesn't find 6.3 or 6.2 when there is a metadata.yml file present.

**Comment by evgeni** [comment]
Given this is `satellite-clone` (and not `katello-clone`), I'd just check the `satellite` RPM and ignore `katello`.

**Comment by johnpmitsch** [comment]
@evgeni updated the module to check the satellite rpm if you can take a look

**Comment by johnpmitsch** [comment]
yee-haw

**Comment by johnpmitsch** [comment]
yee-haw

**Comment by johnpmitsch** [comment]
yee-haw

**Comment by johnpmitsch** [comment]
yee-haw

**Comment by johnpmitsch** [comment]
yee-haw

**Comment by johnpmitsch** [comment]
yee-haw

**Comment by johnpmitsch** [comment]
yee-haw

**Comment by johnpmitsch** [comment]
yee-haw

**Comment by johnpmitsch** [comment]
yee-haw

**Comment by johnpmitsch** [comment]
yee-haw

**Comment by johnpmitsch** [comment]
yee-haw

File: roles/satellite-clone/files/disassociate_capsules.rb [hunk 1]
```diff
@@ -2,7 +2,7 @@
 username = "admin"
 password = "changeme"
 external_capsules = []
-external_capsule_ids = `hammer -u #{username} -p #{password} --csv capsule list --search 'feature = \"Pulp Node\"' | tail -n+2 | awk -F, {'print $1'}`
+external_capsule_ids = `hammer -u #{username} -p #{password} --csv capsule list --search 'feature = \"Pulp Node\"' | grep -v "Warning:" | tail -n+2 | awk -F, {'print $1'}`
```
Comment (sthirugn) [inline roles/satellite-clone/files/disassociate_capsules.rb]: Can you explain why this change was made? this doesn't seem to be related to this PR.

File: roles/satellite-clone/tasks/backup_satellite_version_check.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  register: satellite_answers
+  ignore_errors: true
+  no_log: true
+
+- name: setting fact - satellite_answers
+  set_fact:
+    clone_satellite_answers: "{{ (satellite_answers.rc | int) == 0 }}"
+    cacheable: true
+
+- name: set satellite_version to 6.2 if detected
```
Comment (sthirugn) [inline roles/satellite-clone/tasks/backup_satellite_version_check.yml]: to match the name of the next task perhaps name this `set satellite_version to 6.2 if satellite_answers.yaml exists`

File: roles/satellite-clone/tasks/main.yml [hunk 3]
```diff
@@ -190,17 +169,33 @@
...
   command: katello-installer --capsule-dns false --capsule-dhcp false --capsule-tftp false
-  when: satellite_version == 6.1
-- name: run Satellite 6.2 installer
+  when: satellite_version == "6.1"
+
+- name: "Run Satellite 6.2 installer"
   command: satellite-installer --scenario satellite --foreman-proxy-dns false --foreman-proxy-dhcp false --foreman-proxy-tftp false --foreman-ipa-authentication false
-  when: satellite_version == 6.2
+  environment:
+    HOSTNAME: "{{ hostname }}"
```
Comment (sthirugn) [inline roles/satellite-clone/tasks/main.yml]: what is the need for this?

File: roles/satellite-clone/tasks/main.yml [hunk 4]
```diff
@@ -190,17 +169,33 @@
...
+  when: satellite_version == "6.2"
+
+- name: "Run Satellite 6.3 installer"
+  command: satellite-installer --scenario satellite --foreman-proxy-dns false --foreman-proxy-dhcp false --foreman-proxy-tftp false --foreman-ipa-authentication false --disable-system-checks
```
Comment (sthirugn) [inline roles/satellite-clone/tasks/main.yml]: - what is the purpose of --disable-system checks?
- Is this applicable only for 6.3?

File: roles/satellite-clone/tasks/main.yml [hunk 5]
```diff
@@ -190,17 +169,33 @@
   command: restorecon -R /
   when: restorecon
 
-# Run Satellite installer
-- name: run Satellite 6.1 installer
+# This file tells the candlepin puppet module in the Satellite installer that
+# candlepin is already set up. We remove it so candlepin is set up correctly.
+- name: Remove cpdb_done file
```
Comment (sthirugn) [inline roles/satellite-clone/tasks/main.yml]: did this cause an issue in the past? If so we can create an issue and link to this PR too.

File: roles/satellite-clone/tasks/main.yml [hunk 6]
```diff
@@ -243,20 +240,20 @@
 
   - name: update katello assets
     file: src=/opt/rh/ruby193/root/usr/share/gems/gems/katello-2.2.0.93/public/assets/katello dest=/usr/share/foreman/public/assets/katello
-    when: satellite_version == 6.1
+    when: satellite_version == "6.1"
```
Comment (sthirugn) [inline roles/satellite-clone/tasks/main.yml]: what is the reason behind adding double quotes here?

File: roles/satellite-clone/tasks/restore.yml [hunk 7]
```diff
@@ -6,10 +6,10 @@
   service: name=postgresql state=stopped
 
 - name: Restore postgresql data
-  command: tar --selinux --overwrite -xf {{ backup_dir }}/pgsql_data.tar.gz -C /
+  command: "tar --selinux --overwrite -xf {{ backup_dir }}/pgsql_data.tar.gz -C /"
```
Comment (sthirugn) [inline roles/satellite-clone/tasks/restore.yml]: what is the reason behind including double quotes here?

File: roles/satellite-clone/tasks/backup_satellite_version_check.yml (hunk 2 above)
Comment (johnpmitsch) [reply roles/satellite-clone/tasks/backup_satellite_version_check.yml]: Thanks, since it will be updated to 6.3 if its a 6.3 backup, how about we say `Temporarily set satellite_version to 6.2 if satellite_answers.yaml exists`

File: roles/satellite-clone/tasks/main.yml (hunk 3 above)
Comment (johnpmitsch) [reply roles/satellite-clone/tasks/main.yml]: So this is something that only occurs in 6.3, but I added in 6.2 as well to be safe. 

The $HOSTNAME bash variable doesn't update when running the playbook after we change the hostname. Some of the puppet modules that the installer runs rely on this bash variable. In this case it was making calls to the foreman-proxy using $HOSTNAME, but using the old $HOSTNAME bash variable, before it was changed to match the backup's hostname. This ensures that the bash variable is updated.

File: roles/satellite-clone/tasks/main.yml (hunk 4 above)
Comment (johnpmitsch) [reply roles/satellite-clone/tasks/main.yml]: Yes, I was confused about this as well, I guess it is not an option in 6.2 (I thought it was...) but the RAM check is going to block jenkins. I figure if someone has already installed and backed up Satellite on one machine (which they have if they are running clone), they understand the system requirements needed to run Satellite and we shouldn't be checking this as part of sat-clone and failing playbooks because of the system requirements check.

File: roles/satellite-clone/tasks/main.yml (hunk 5 above)
Comment (johnpmitsch) [reply roles/satellite-clone/tasks/main.yml]: I don't think it has in the past, but it definitely is causing an issue in 6.3.

File: roles/satellite-clone/tasks/main.yml (hunk 6 above)
Comment (johnpmitsch) [reply roles/satellite-clone/tasks/main.yml]: So the check_metadata_for_satellite_version module returns a string and I can't get it to return an integer. I'm guessing this has to do with how `register` works in ansible. I changed satellite_version to always be a string to match the new module and updated usage appropiately

File: roles/satellite-clone/tasks/restore.yml (hunk 7 above)
Comment (johnpmitsch) [reply roles/satellite-clone/tasks/restore.yml]: It was part of debugging a different issue (that was solved by the delete `cp_done` task) but I think we actually should be using quotes wherever string interpolation `{{ string }}` is used

File: roles/satellite-clone/files/disassociate_capsules.rb (hunk 1 above)
Comment (johnpmitsch) [reply roles/satellite-clone/files/disassociate_capsules.rb]: So on 6.3 hammer is displaying a warning for the first line of the hammer command output. The `tail` command here is expecting everything after the second line, which makes sense for how the normal hammer csv output is, but the warning is bumping everything down a line. I added this to remove any lines containing `Warning:` so the count remains correct.

**Comment by sthirugn** [comment]
Testing in progress:
- [x] 6.2 clone test - PASS
- [ ] 6.3 clone test
- [ ] 6.4 clone test

**Comment by johnpmitsch** [comment]
@evgeni friendly poke, can you take another glance over the code? :pray:

File: library/check_metadata_for_satellite_version.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+def check_metatadata_for_backup(params):
+    with open(params["metadata_path"]) as data_file:
+        data = yaml.load(data_file)
+
+    rpms = data[":rpms"]
+    pattern = re.compile("^satellite-[\d+].*")
+    satellite = [r for r in rpms if pattern.match(r)][0]
+    satellite_version = satellite.split("-")[1]
+
+    if LooseVersion(satellite_version) >= LooseVersion("6.2.0") and \
```
Comment (evgeni) [inline library/check_metadata_for_satellite_version.py]: Is there a reason not to just do
```python
found_version = '.'.join(satellite_version.split('.')[0:1])
```

File: library/check_metadata_for_satellite_version.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+#    - Return the Satellite version specified in a Satellite backup
+# notes:
+#    - The Satellite version is determined from the Satellite rpm
+#      version using the backup's rpm list from metadata.yml
+# options:
+#    metadata_path:
+#        description:
+#          - Full path (including file name) to metadata.yml
+#        required: true
+def check_metatadata_for_backup(params):
```
Comment (evgeni) [inline library/check_metadata_for_satellite_version.py]: I'd argue this is not a check, but `get_satellite_backup_version`? :)

File: roles/satellite-clone/tasks/main.yml (hunk 5 above)
Comment (evgeni) [reply roles/satellite-clone/tasks/main.yml]: but, given we're restoring a working backup, shouldn't candlepin actually be properly setup at this point?

File: roles/satellite-clone/tasks/main.yml (hunk 3 above)
Comment (evgeni) [reply roles/satellite-clone/tasks/main.yml]: I wonder if [`meta: reset_connection`](https://docs.ansible.com/ansible/latest/meta_module.html) would help here and be less of a hack

**Comment by evgeni** [comment]
@johnpmitsch added some comments, nothing major

File: roles/satellite-clone/tasks/main.yml [hunk 10]
```diff
@@ -229,11 +224,13 @@
...
     command: foreman-rake foreman_tasks:cleanup TASK_SEARCH='label ~ *' STATES='paused'  VERBOSE=true AFTER='0h'  VERBOSE=true
-    when: satellite_version == 6.2
+    when: satellite_version in ["6.2", "6.3"]
 
-  - name: Run installer upgrade (satellite 6.2 only)
+  - name: Run installer upgrade (satellite 6.2+ only)
     command: satellite-installer --upgrade
-    when: satellite_version == 6.2
+    environment:
+      HOSTNAME: "{{ hostname }}"
```
[the remark this answers is not in the document]
Comment (johnpmitsch) [reply roles/satellite-clone/tasks/main.yml]: meta comment?

[the remark this answers is not in the document]
Comment (evgeni) [reply roles/satellite-clone/tasks/main.yml]: https://github.com/RedHatSatellite/satellite-clone/pull/325#discussion_r174420139

[the remark this answers is not in the document]
Comment (johnpmitsch) [reply roles/satellite-clone/tasks/main.yml]: missed that one, thanks!

File: roles/satellite-clone/tasks/main.yml (hunk 5 above)
Comment (johnpmitsch) [reply roles/satellite-clone/tasks/main.yml]: not exactly, here are the rough steps of clone and what is happening
- restore configs  <---- cpdb_done file is restored from the backup
- install Satellite <----- installer sees cpdb_done file and says "ok I wont set up candlepin then", resulting in errors on seed steps when installing
- restore backups (mongo, postgres, pulp data)
- re-run installer

**Comment by johnpmitsch** [comment]
@evgeni updated

**Comment by johnpmitsch** [comment]
yee-haw

**Comment by sthirugn** [comment]
Update satellite-clone.md to reflect the fact that metadata.yml is mandatory for 6.3 and above

**Comment by sthirugn** [comment]
updated test status in comment https://github.com/RedHatSatellite/satellite-clone/pull/325#issuecomment-372759142

**Comment by johnpmitsch** [comment]
yee-haw

**Comment by sthirugn** [comment]
ACK

**Comment by johnpmitsch** [comment]
yee-haw

File: roles/satellite-clone/files/disassociate_capsules.rb (hunk 1 above)
Comment (evgeni) [reply roles/satellite-clone/files/disassociate_capsules.rb]: Which warning, is there a BZ about that? :-)

File: roles/satellite-clone/tasks/main.yml (hunk 5 above)
Comment (evgeni) [reply roles/satellite-clone/tasks/main.yml]: Aah, gotcha. I'd argue there should be an RFE to the installer, to force Candlepin re-init (like we can with certs), but that remove will do for now.

**Comment by johnpmitsch** [comment]
@evgeni good point, here is [hammer BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1556948) and [cpdb_done RFE for katello](http://projects.theforeman.org/issues/22913)
